cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -280,6 +307,88 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
+template <typename FPType, CpuType cpu>
+class CosineDistances : public EuclideanDistances<FPType, cpu>
+{
+public:
+    using super = EuclideanDistances<FPType, cpu>;
+
+    CosineDistances(const NumericTable & a, const NumericTable & b) : super(a, b, true) {}
+
+    virtual ~CosineDistances() {}
+
+    PairwiseDistanceType getType() override { return PairwiseDistanceType::cosine; }
+
+    services::Status finalize(const size_t n, FPType * a) override { return services::Status(); }
+
+    // output:  Row-major matrix of size { aSize x bSize }
+    services::Status computeBatch(const FPType * const a, const FPType * const b, size_t aOffset, size_t aSize, size_t bOffset, size_t bSize,
+                                  FPType * const res) override
+    {
+        const size_t nRowsA = aSize;
+        const size_t nColsA = super::_a.getNumberOfColumns();
+        const size_t nRowsB = bSize;
+
+        const size_t nRowsC = nRowsA;
+        const size_t nColsC = nRowsB;
+
+        super::computeABt(a, b, nRowsA, nColsA, nRowsB, res);
+
+        const FPType * const aa = super::normBufferA.get() + aOffset;
+        const FPType * const bb = (&(super::_a) == &(super::_b)) ? super::normBufferA.get() + bOffset : super::normBufferB.get() + bOffset;
+
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t i = 0; i < nRowsC; i++)
+        {
+            for (size_t j = 0; j < nColsC; j++)
+            {
+                res[i * nColsC + j] = FPType(1) - (res[i * nColsC + j] / (aa[i] * bb[j]));
+            }
+        }
+        return services::Status();
+    }
+
+protected:
+    services::Status computeNorm(const NumericTable & ntData, FPType * const res) override
+    {
+        const size_t nRows = ntData.getNumberOfRows();
+        const size_t nCols = ntData.getNumberOfColumns();
+
+        const size_t blockSize = 512;
+        const size_t nBlocks   = nRows / blockSize + !!(nRows % blockSize);
+
+        SafeStatus safeStat;
+        daal::threader_for(nBlocks, nBlocks, [&](size_t iBlock) {
+            const size_t begin         = iBlock * blockSize;
+            const size_t end           = services::internal::min<cpu, size_t>(begin + blockSize, nRows);
+            const size_t blockSizeReal = end - begin;
+
+            ReadRows<FPType, cpu> dataRows(const_cast<NumericTable &>(ntData), begin, blockSizeReal);
+            DAAL_CHECK_BLOCK_STATUS_THR(dataRows);
+            const FPType * const data = dataRows.get();
+
+            FPType * r = res + begin;
+
+            for (size_t i = 0; i < blockSizeReal; i++)
+            {
+                FPType sum = FPType(0);
+                PRAGMA_IVDEP
+                PRAGMA_ICC_NO16(omp simd reduction(+ : sum))
+                for (size_t j = 0; j < nCols; j++)
+                {
+                    sum += data[i * nCols + j] * data[i * nCols + j];
+                }
+                r[i] = sum;
+            }
+            Math<FPType, cpu>::vSqrt(blockSizeReal, r, r);
```

Comment:
So, only one difference between computeNorm for Euclidean and Cosine - this Sqrt?
If so, better to extract common code somehow, even ```if``` is better